### PR TITLE
Avoid copying regions of the sentence buffer

### PR DESCRIPTION
**Pull request by guillaumekln**

### CODE REVIEW

File: src/ngram_matches.cc [hunk 1]
```diff
@@ -44,14 +44,13 @@ namespace fuzzy
     return _psentences.size();
   }
 
-  std::vector<unsigned>
-  NGramMatches::sentence(int s_id) const
+  const unsigned*
+  NGramMatches::sentence(unsigned s_id, unsigned* length) const
   {
     unsigned idx = _suffixArray[s_id];
```
Comment (ClementChouteau) [inline src/ngram_matches.cc]: Automatic type deduction should be fine here, I suggest replacing type with `const auto`, and line below with `const auto*`.

File: src/ngram_matches.cc [hunk 2]
```diff
@@ -44,14 +44,13 @@ namespace fuzzy
     return _psentences.size();
   }
 
-  std::vector<unsigned>
-  NGramMatches::sentence(int s_id) const
+  const unsigned*
+  NGramMatches::sentence(unsigned s_id, unsigned* length) const
```
Comment (ClementChouteau) [inline src/ngram_matches.cc]: Maybe you should introduce a `SuffixString` (or maybe with a better name), to group the return value and `length`.

File: src/fuzzy_match.cc [hunk 3]
```diff
@@ -579,12 +582,13 @@ namespace fuzzy
...
       {
         Costs costs;
-        costs.diff_word = 100. / std::max(suffix_wids.size(), p_length);
+        costs.diff_word = 100. / std::max(s_length, p_length);
 
         /* let us check the candidates */
         const auto suffix_realtok = _suffixArrayIndex->real_tokens(s_id);
         float cost = _edit_distance(suffix_wids, suffix_realtok,
-                                    pattern_wids, pattern_realtok,
+                                    s_length,
```
Comment (ClementChouteau) [inline src/fuzzy_match.cc:589]: I suggest putting `suffix_wids, suffix_realtok, s_length` on the same line, and `pattern_wids.data(), pattern_realtok, p_length` on another to better group related parameters.

File: src/fuzzy_match.cc [hunk 4]
```diff
@@ -552,13 +552,16 @@ namespace fuzzy
     for (auto agendaItemIt = nGramMatches.get_psentences().begin(); agendaItemIt != nGramMatches.get_psentences().end(); ++agendaItemIt)
     {
       auto& agendaItem = agendaItemIt.value();
-      int s_id = agendaItem.s_id;
-      const auto suffix_wids = nGramMatches.sentence(s_id); //TODO we may have to update this
+      unsigned s_id = agendaItem.s_id;
```
Comment (ClementChouteau) [inline src/fuzzy_match.cc]: Use `auto`

**Comment by ClementChouteau** [comment]
@guillaumekln It looks good, please check the tests with Valgrind.

**Comment by guillaumekln** [comment]
Thanks. Valgrind reports no error.

**Pull request merged**